company-lsp.el

```diff
@@ -272,12 +272,37 @@ Return a list of strings as the completion candidates."
   (remove-hook 'company-completion-finished-hook #'company-lsp--cleanup-cache)
   (remove-hook 'company-completion-cancelled-hook #'company-lsp--cleanup-cache))
 
+(defun company-lsp--cache-put (prefix candidates)
+  "Set cache for PREFIX to be CANDIDATES.
+
+CANDIDATES is a plist of (:incomplete :candidates). :incomplete is
+either t or nil. :candidates is a list of strings."
+  (setq company-lsp--completion-cache
+        (cons (cons prefix candidates)
+              company-lsp--completion-cache)))
+
 (defun company-lsp--cache-get (prefix)
   "Get the cached completion for PREFIX.
 
 Return a plist of (:incomplete :candidates) if cache for PREFIX
 exists. Otherwise return nil."
-  (cdr (assoc prefix company-lsp--completion-cache)))
+  (let ((cache (cdr (assoc prefix company-lsp--completion-cache)))
+        (len (length prefix))
+        previous-cache)
+    (if cache
+        cache
+      (cl-dotimes (i (1+ len))
```

Comment:
It is correct `(1+ len)`? I would expect just `len`.
Otherwise below we can get `(- len (1+ len))`, which would be `-1` as the 3rd argument to substring, no?

Reply:
`cl-dotimes` loops `i` in the range of `[0,  1 + len)`, so the last value is `len`, not `1 + len`. It's meaningless when `i = 0` though. I changed the loop to `[0, len)` and get the substring for `[0, len - i - 1)` instead.